geoserve/geoserve.go

```diff
@@ -172,45 +163,66 @@ func readDbFromFile(dbFile string) (*geoip2.Reader, time.Time, error) {
 	if err != nil {
 		return nil, time.Time{}, fmt.Errorf("Unable to stat db file %s: %s", dbFile, err)
 	}
-	dbDate := fileInfo.ModTime()
+	lastModified := fileInfo.ModTime()
 	db, err := openDb(dbData)
 	if err != nil {
 		return nil, time.Time{}, err
 	} else {
-		return db, dbDate, nil
+		return db, lastModified, nil
 	}
 }
 
 // readDbFromWeb reads the MaxMind database and timestamp from the web
-func readDbFromWeb() (*geoip2.Reader, time.Time, error) {
-	dbResp, err := http.Get(DB_URL)
+func readDbFromWeb(url string, ifModifiedSince time.Time) (*geoip2.Reader, time.Time, error) {
+	req, err := http.NewRequest(http.MethodGet, url, nil)
 	if err != nil {
-		return nil, time.Time{}, fmt.Errorf("Unable to get database from %s: %s", DB_URL, err)
+		return nil, time.Time{}, err
 	}
-	gzipDbData, err := gzip.NewReader(dbResp.Body)
+	req.Header.Add("If-Modified-Since", ifModifiedSince.Format(http.TimeFormat))
+	resp, err := http.DefaultClient.Do(req)
 	if err != nil {
-		return nil, time.Time{}, fmt.Errorf("Unable to open gzip reader on response body%s", err)
+		return nil, time.Time{}, fmt.Errorf("Unable to get database from %s: %s", url, err)
+	}
+	if resp.StatusCode == http.StatusNotModified {
+		return nil, time.Time{}, errNotModified
 	}
-	defer gzipDbData.Close()
-	dbData, err := ioutil.ReadAll(gzipDbData)
+	if resp.StatusCode != http.StatusOK {
+		return nil, time.Time{}, fmt.Errorf("unexpected HTTP status %v", resp.Status)
+	}
+	lastModified, err := getLastModified(resp)
 	if err != nil {
-		return nil, time.Time{}, fmt.Errorf("Unable to fetch database from HTTP response: %s", err)
+		return nil, time.Time{}, fmt.Errorf("Unable to parse Last-Modified header %s: %s", lastModified, err)
 	}
-	dbDate, err := lastModified(dbResp)
+	defer resp.Body.Close()
```

Comment:
I think as long as the call to http client `Do` on line 182 didn't return an error, it's safe to call resp.Body.Close() so maybe you could move this defer to line 186 to make sure we're closing the body in all cases?

Reply:
Yeah sure, will do.